man/man4/tcp.4
.\" Copyright (c) 1983 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms are permitted
.\" provided that the above copyright notice and this paragraph are
.\" duplicated in all such forms and that any documentation,
.\" advertising materials, and other materials related to such
.\" distribution and use acknowledge that the software was developed
.\" by the University of California, Berkeley.  The name of the
.\" University may not be used to endorse or promote products derived
.\" from this software without specific prior written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\"	@(#)tcp.4	6.3 (Berkeley) 2/14/89
.\"
.TH TCP 4 "February 14, 1989"
.UC 5
.SH NAME
tcp \- Internet Transmission Control Protocol
.SH SYNOPSIS
.B #include <sys/socket.h>
.br
.B #include <netinet/in.h>
.PP
.B s = socket(AF_INET, SOCK_STREAM, 0);
.SH DESCRIPTION
The TCP protocol provides reliable, flow-controlled, two-way
transmission of data.  It is a byte-stream protocol used to
support the SOCK_STREAM abstraction.  TCP uses the standard
Internet address format and, in addition, provides a per-host
collection of \*(lqport addresses\*(rq.  Thus, each address is composed
of an Internet address specifying the host and network, with
a specific TCP port on the host identifying the peer entity.
.PP
Sockets utilizing the tcp protocol are either \*(lqactive\*(rq or
\*(lqpassive\*(rq.  Active sockets initiate connections to passive
sockets.  By default TCP sockets are created active; to create a
passive socket the
.IR listen (2)
system call must be used
after binding the socket with the
.IR bind (2)
system call.  Only
passive sockets may use the 
.IR accept (2)
call to accept incoming connections.  Only active sockets may
use the
.IR connect (2)
call to initiate connections.
.PP
Passive sockets may \*(lqunderspecify\*(rq their location to match
incoming connection requests from multiple networks.  This
technique, termed \*(lqwildcard addressing\*(rq, allows a single
server to provide service to clients on multiple networks.
To create a socket which listens on all networks, the Internet
address INADDR_ANY
must be bound.  The TCP port may still be specified
at this time; if the port is not specified the system will assign one.
Once a connection has been established the socket's address is
fixed by the peer entity's location.   The address assigned the
socket is the address associated with the network interface
through which packets are being transmitted and received.  Normally
this address corresponds to the peer entity's network.
.PP
TCP supports one socket option which is set with
.IR setsockopt (2)
and tested with
.IR getsockopt (2).
Under most circumstances, TCP sends data when it is presented;
when outstanding data has not yet been acknowledged, it gathers
small amounts of output to be sent in a single packet once
an acknowledgement is received.
For a small number of clients, such as window systems
that send a stream of mouse events which receive no replies,
this packetization may cause significant delays.
Therefore, TCP provides a boolean option, TCP_NODELAY (from
.IR <netinet/tcp.h> ,
to defeat this algorithm.
The option level for the
.I setsockopt
call is the protocol number for TCP,
available from
.IR getprotobyname (3).
.PP
Options at the IP transport level may be used with TCP; see
.IR ip (4).
Incoming connection requests that are source-routed are noted,
and the reverse source route is used in responding.
.SH DIAGNOSTICS
A socket operation may fail with one of the following errors returned:
.TP 20
[EISCONN]
when trying to establish a connection on a socket which
already has one;
.TP 20
[ENOBUFS]
when the system runs out of memory for
an internal data structure;
.TP 20
[ETIMEDOUT]
when a connection was dropped
due to excessive retransmissions;
.TP 20
[ECONNRESET]
when the remote peer
forces the connection to be closed;
.TP 20
[ECONNREFUSED]
when the remote
peer actively refuses connection establishment (usually because
no process is listening to the port);
.TP 20
[EADDRINUSE]
when an attempt
is made to create a socket with a port which has already been
allocated;
.TP 20
[EADDRNOTAVAIL]
when an attempt is made to create a 
socket with a network address for which no network interface
exists.
.SH SEE ALSO
getsockopt(2), socket(2), intro(4), inet(4), ip(4)


named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


.ref-386BSD-0.1-patchkit/usr/src/libexec/uucp/tcp.c
/* tcp.c
   Code to handle TCP connections.

   Copyright (C) 1991, 1992 Ian Lance Taylor

   This file is part of the Taylor UUCP package.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

   The author of the program may be contacted at ian@airs.com or
   c/o AIRS, P.O. Box 520, Waltham, MA 02254.

   $Log: tcp.c,v $
   Revision 1.16  1992/03/30  15:37:22  ian
   Petri Helenius: TCP server never started uuxqt

   Revision 1.15  1992/03/17  02:44:08  ian
   Declare _exit

   Revision 1.14  1992/03/17  00:34:10  ian
   Cast argument to bzero

   Revision 1.13  1992/03/15  04:51:17  ian
   Keep an array of signals we've received rather than a single variable

   Revision 1.12  1992/03/15  01:54:46  ian
   All execs are now done in isspawn, all waits are done in iswait

   Revision 1.11  1992/03/12  19:56:10  ian
   Debugging based on types rather than number

   Revision 1.10  1992/03/11  02:09:57  ian
   Franc,ois Pinard: retry fork several times before giving up

   Revision 1.9  1992/03/10  22:57:03  ian
   Petri Helenius: have server fork twice to avoid zombies

   Revision 1.8  1992/02/08  03:54:18  ian
   Include <string.h> only in <uucp.h>, added 1992 copyright

   Revision 1.7  1991/12/28  03:49:23  ian
   Added HAVE_MEMFNS and HAVE_BFNS; changed uses of memset to bzero

   Revision 1.6  1991/12/26  22:54:32  ian
   Monty Solomon: cast arguments to avoid prototype errors

   Revision 1.5  1991/12/22  22:14:19  ian
   Monty Solomon: added HAVE_UNISTD_H configuration parameter

   Revision 1.4  1991/12/10  19:45:05  ian
   Added ulog_device to record device name for log file

   Revision 1.3  1991/11/24  20:06:08  ian
   Michael Haberler: itcp_port_number of a number wasn't calling htons

   Revision 1.2  1991/11/14  03:20:13  ian
   Added seven-bit and reliable commands to help when selecting protocols

   Revision 1.1  1991/11/13  20:38:00  ian
   Initial revision

   */

#include "uucp.h"

#if USE_RCS_ID
char tcp_rcsid[] = "$Id: tcp.c,v 1.16 1992/03/30 15:37:22 ian Rel $";
#endif

#if HAVE_TCP

#include <errno.h>

#if USE_STDIO && HAVE_UNISTD_H
#include <unistd.h>
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>

#include "sysdep.h"
#include "port.h"
#include "system.h"

/* External functions.  */
extern int close ();
extern int socket (), bind (), listen (), accept (), connect ();
extern void _exit ();

/* This code handles TCP connections by providing a set of routines
   that are called by the port routines.  It assumes a Berkeley socket
   interface.  */

/* The normal "uucp" port number.  */
#define IUUCP_PORT (540)

/* Local function to get the port number from a name.  */
static int itcp_port_number P((const char *zport));

/* Lock a port.  There is no need to lock a TCP port.  */

boolean
ftcp_lock (qport, fin)
     struct sport *qport;
     boolean fin;
{
  return TRUE;
}

/* Open a port.  If the fwait argument is TRUE, we are running as a
   server.  Otherwise we are just trying to reach another system.  */

boolean
ftcp_open (qport, ibaud, fwait)
     struct sport *qport;
     long ibaud;
     boolean fwait;
{
  struct sockaddr_in s;

  ulog_device ("TCP");

  qport->u.stcp.o = socket (AF_INET, SOCK_STREAM, 0);
  if (qport->u.stcp.o == -1)
    {
      ulog (LOG_ERROR, "socket: %s", strerror (errno));
      return FALSE;
    }

  /* If we aren't waiting for a connection, we're done.  */

  if (! fwait)
    return TRUE;

  /* Run as a server and wait for a new connection.  The code in
     uucico.c has already detached us from our controlling terminal.
     From this point on if the server gets an error we exit; we only
     return if we have received a connection.  It would be more robust
     to respawn the server if it fails; someday.  */

  bzero ((pointer) &s, sizeof s);
  s.sin_family = AF_INET;
  s.sin_port = itcp_port_number (qport->u.stcp.zport);
  s.sin_addr.s_addr = htonl (INADDR_ANY);

  if (bind (qport->u.stcp.o, (struct sockaddr *) &s, sizeof s) == -1)
    {
      ulog (LOG_ERROR, "bind: %s", strerror (errno));
      usysdep_exit (FALSE);
    }
  if (listen (qport->u.stcp.o, 5) == -1)
    {
      ulog (LOG_ERROR, "listen: %s", strerror (errno));
      usysdep_exit (FALSE);
    }

  while (! FGOT_SIGNAL ())
    {
      int clen;
      int onew;
      pid_t ipid;

      DEBUG_MESSAGE0 (DEBUG_PORT,
		      "ftcp_open: Waiting for connections");

      clen = sizeof s;
      onew = accept (qport->u.stcp.o, (struct sockaddr *) &s, &clen);
      if (onew == -1)
	{
	  ulog (LOG_ERROR, "accept: %s", strerror (errno));
	  usysdep_exit (FALSE);
	}

      DEBUG_MESSAGE0 (DEBUG_PORT,
		      "ftcp_open: Got connection; forking");

      ipid = isfork ();
      if (ipid < 0)
	{
	  ulog (LOG_ERROR, "fork: %s", strerror (errno));
	  usysdep_exit (FALSE);
	}
      if (ipid == 0)
	{
	  (void) close (qport->u.stcp.o);
	  qport->u.stcp.o = onew;

	  /* Now we fork and let our parent die, so that we become
	     a child of init.  This lets the main server code wait
	     for its child and then continue without accumulating
	     zombie children.  */

	  ipid = isfork ();
	  if (ipid < 0)
	    {
	      ulog (LOG_ERROR, "fork: %s", strerror (errno));
	      _exit (EXIT_FAILURE);
	    }
	      
	  if (ipid != 0)
	    _exit (EXIT_SUCCESS);

	  return TRUE;
	}

      (void) close (onew);

      /* Now wait for the child.  */
      (void) iswait ((unsigned long) ipid, (const char *) NULL);
    }

  /* We got a signal.  */
  usysdep_exit (FALSE);

  /* Avoid compiler warnings.  */
  return FALSE;
}

/* Close the port.  */

boolean
ftcp_close (qport, fsuccess)
     struct sport *qport;
     boolean fsuccess;
{
  if (close (qport->u.stcp.o) == -1)
    {
      ulog (LOG_ERROR, "close: %s", strerror (errno));
      return FALSE;
    }
  return TRUE;
}

/* Reset the port.  This will be called by a child which was forked
   off in ftcp_open, above.  We don't want uucico to continue looping
   and giving login prompts, so we pretend that we received a SIGINT
   signal.  This should probably be handled more cleanly.  */

boolean
ftcp_reset (qport)
     struct sport *qport;
{
  afSignal[INDEXSIG_SIGINT] = TRUE;
  return TRUE;
}

/* Dial out on a TCP port, so to speak: connect to a remote computer.  */

boolean
ftcp_dial (qport, qsys, pcproto_params, pqproto_params, pireliable)
     struct sport *qport;
     const struct ssysteminfo *qsys;
     int *pcproto_params;
     struct sproto_param **pqproto_params;
     int *pireliable;
{
  const char *zhost;
  struct sockaddr_in s;
  struct hostent *q;

  /* There are no dialer protocol parameters, but a TCP connection is
     fully reliable.  */
  *pcproto_params = 0;
  *pqproto_params = NULL;
  *pireliable = (RELIABLE_SPECIFIED | RELIABLE_ENDTOEND
		 | RELIABLE_RELIABLE | RELIABLE_EIGHT);

  zhost = qsys->zphone;
  if (zhost == NULL)
    zhost = qsys->zname;

  /* Cast argument to avoid prototype error on NeXT.  */
  q = gethostbyname ((char *) zhost);
  if (q == NULL)
    {
      ulog (LOG_ERROR, "gethostbyname (%s): %s", zhost, strerror (errno));
      return FALSE;
    }

  s.sin_family = q->h_addrtype;
  s.sin_port = itcp_port_number (qport->u.stcp.zport);
  memcpy (&s.sin_addr.s_addr, q->h_addr, q->h_length);

  if (connect (qport->u.stcp.o, (struct sockaddr *) &s, sizeof s) == -1)
    {
      ulog (LOG_ERROR, "connect: %s", strerror (errno));
      return FALSE;
    }

  return TRUE;
}

/* Return the baud rate.  */

long
itcp_baud (qport)
     struct sport *qport;
{
  return 0;
}

/* Get the port number given a name.  The argument will almost always
   be "uucp" so we cache that value.  The return value is always in
   network byte order.  This returns -1 on error.  */

static int
itcp_port_number (zname)
     const char *zname;
{
  boolean fuucp;
  static int iuucp;
  int i;
  char *zend;
  struct servent *q;

  fuucp = strcmp (zname, "uucp") == 0;
  if (fuucp && iuucp != 0)
    return iuucp;

  /* Try it as a number first.  */
  i = strtol (zname, &zend, 10);
  if (i != 0 && *zend == '\0')
    return htons (i);

  /* Cast arguments to avoid prototype errors on NeXT.  */
  q = getservbyname ((char *) zname, (char *) "tcp");
  if (q == NULL)
    {
      /* We know that the "uucp" service should be 540, even if isn't
	 in /etc/services.  */
      if (fuucp)
	{
	  iuucp = htons (IUUCP_PORT);
	  return iuucp;
	}
      ulog (LOG_ERROR, "getservbyname (%s): %s", zname, strerror (errno));
      return -1;
    }

  if (fuucp)
    iuucp = q->s_port;

  return q->s_port;
}

#endif /* HAVE_TCP */
